[GStreamer][MediaStream] fast/mediastream/MediaStream-video-element-track-stop.html fails

philn opened the pull request:
#### d42fe538c8b6f8f01d268722793c4420a4573abf
<pre>
[GStreamer][MediaStream] fast/mediastream/MediaStream-video-element-track-stop.html fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=254529">https://bugs.webkit.org/show_bug.cgi?id=254529</a>

Reviewed by Xabier Rodriguez-Calvar.

The EOS and position tracking for MediaStreams is improved, we now report monotonically increasing
timestamps for the position, similarily to the Mac port. The duration is no longer cached, it&apos;s
useless in this case, we can always return positiveInfinite() and thus simplify the `didEnd()` method.

Also the stream collection is now handled differently, when receiving a new collection we
first disable video/text tracks, remove old audio tracks and reactivate tracks corresponding to
existing GstStreams. When player no longer has enabled tracks we trigger the EOS mechanism.

The mediastreamsrc element will now post an empty stream collection when all its tracks have ended,
this allows the media player to deactivate ended video tracks and remove ended audio tracks.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::play):
(WebCore::MediaPlayerPrivateGStreamer::pause):
(WebCore::MediaPlayerPrivateGStreamer::durationMediaTime const):
(WebCore::MediaPlayerPrivateGStreamer::currentMediaTime const):
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
(WebCore::MediaPlayerPrivateGStreamer::didEnd):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(webkitMediaStreamSrcPostStreamCollection):

Canonical link: <a href="https://commits.webkit.org/263307@main">https://commits.webkit.org/263307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf928411eea3061cc82a58f3d4ab533a66325f8f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4208 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5676 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4454 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4200 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4291 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/4675 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4444 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5669 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1939 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3772 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/5956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3766 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3838 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4246 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3446 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3757 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3770 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1032 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4094 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2621,18 +2659,20 @@ void MediaPlayerPrivateGStreamer::didEnd()
+        // Now that playback has ended it's NOT a safe time to send a SELECT_STREAMS event. In fact, as of GStreamer 1.16,
+        // playbin3 will crash on a GStreamer assertion (combine->sinkpad being unexpectedly null) if we try. Instead, wait
+        // until we get the initial STREAMS_SELECTED message one more time.
+        m_waitingForStreamsSelectedEvent = true;
```
Why is this indented?

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -610,6 +611,10 @@ class MediaPlayerPrivateGStreamer : public MediaPlayerPrivateInterface
+
+    // Specific to MediaStream playback.
+    MediaTime m_startTime;
+    MediaTime m_pausedTime;
```
Shouldn't we flag this for `MEDIA_TRACK`?

philn replied:
`MEDIA_STREAM` you mean? Yes, it wouldn't hurt, I guess :)

philn replied:
Actually it would complicate even more code with ifdefs. Doesn't seem worth it.

philn committed:
[GStreamer][MediaStream] fast/mediastream/MediaStream-video-element-track-stop.html fails
https://bugs.webkit.org/show_bug.cgi?id=254529

Reviewed by Xabier Rodriguez-Calvar.

The EOS and position tracking for MediaStreams is improved, we now report monotonically increasing
timestamps for the position, similarily to the Mac port. The duration is no longer cached, it's
useless in this case, we can always return positiveInfinite() and thus simplify the `didEnd()` method.

Also the stream collection is now handled differently, when receiving a new collection we
first disable video/text tracks, remove old audio tracks and reactivate tracks corresponding to
existing GstStreams. When player no longer has enabled tracks we trigger the EOS mechanism.

The mediastreamsrc element will now post an empty stream collection when all its tracks have ended,
this allows the media player to deactivate ended video tracks and remove ended audio tracks.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::play):
(WebCore::MediaPlayerPrivateGStreamer::pause):
(WebCore::MediaPlayerPrivateGStreamer::durationMediaTime const):
(WebCore::MediaPlayerPrivateGStreamer::currentMediaTime const):
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
(WebCore::MediaPlayerPrivateGStreamer::didEnd):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(webkitMediaStreamSrcPostStreamCollection):

Canonical link: https://commits.webkit.org/263307@main
